Maint/cpp 11 cleanup

JWhitleyWork opened the pull request:
Includes the following:

1. Add the run_tests hook for roslint.
2. Transfers data as std::vectors instead of raw pointers to arrays.
3. Creates a `common.h` header with enums that are used in `network_interface.h` and `network_utils.h`.
4. Makes all enums scoped.
5. In network_utils, uses the size of the requested type for reading/writing instead of an argument.
6. In network_interface, uses the size of the provided vector for writing and allocates a large buffer (10,000 bytes) for reading. Number of actual bytes read is still returned on read.
7. Removes unnecessary .orig file.
8. Removes no-longer-necessary utils.cpp and moves the single function from that file into `network_utils.h`.

Commit:
CI: Making roslint run in run_tests.

Commit:
Using vectors and type traits.

Commit:
Scoped enum for return statuses.

Commit:
Moving common enums into common.h.

Commit:
Switching back to memcpy.

Commit:
Making ByteOrder enum scoped too.

Commit:
Reserving requested space in read_exactly.

JWhitleyWork commented:
Addresses #14.

JWhitleyWork commented:
Also, the approval of this PR will impact https://github.com/astuff/ibeo_scala and https://github.com/astuff/ibeo_lux. Once this PR has been hashed out and is approved, we should leave it un-merged until the accompanying PRs are ready in those repos.

Also, all the more reason for #23.

Commit:
-DBUILD_ROS is no longer a thing.

zoakesAStuff commented on include/network_interface/network_utils.h:
```diff
@@ -77,14 +78,14 @@ std::vector<uint8_t> write_le(T *source)
   if (sizeof(source))
```
IIRC sizeof always returns at least 1, even for empty structures/classes. This seems a little fishy to me. See https://en.cppreference.com/w/cpp/language/sizeof

JWhitleyWork replied:
Yeah, I agree. Very smelly code. Not sure what the original author was checking for here but I think we can just remove this check.

zoakesAStuff commented on src/tcp_interface.cpp:
```diff
@@ -107,8 +100,9 @@ return_statuses TCPInterface::read(unsigned char *msg,
                                  boost::asio::placeholders::error));
   }
 
+  msg->assign(10000, 0);
```
Assigning a huge buffer to a vector and then treating it as a buffer is fine, but might be better to use vector::reserve and then vector:shrink_to_fit. That way there's less unneeded memory that stays allocated. Either way you'll have to manage msg.size manually since this call to assign sets msg.size to 10000.

JWhitleyWork replied:
Since shrink_to_fit just shrinks the allocated memory to fit the size and reserve doesn't increase the size, this will still result in a vector with size == 0 and 0 memory reserved. Allocation through assign, push_back, or emplace are the only ways to make the vector size actually grow. Since the actual number of bytes read is returned in this function, I figured I would just assign something huge and the user will know how many bytes are valid through a read. I'm open to other suggestions.

zoakesAStuff replied:
You're right. It'd be better to call vector::resize(bytes_read) instead of vector::shrink_to_size(). If we're not going to actually use the dynamic allocation, why are we using vectors?

JWhitleyWork replied:
But you can't because it needs the buffer before the bytes have been read. Let me look more into how boost::buffer with a vector works. It might do the allocation itself and this may not be necessary.

JWhitleyWork replied:
So, I'm reading over this code more and I think this call to async_read is a terrible idea as implemented. I think I did it because you can only provide a timeout when doing an async read but doing it as implemented means that the vector that gets passed in is managed by the API user and they would have to know that they can't modify the vector until the read returns, which isn't currently being triggered. I need to rethink this and implement it differently. I'm going to cancel this PR until that is done.

JWhitleyWork replied:
Used reserve() instead and the boost::buffer allocates the bytes as necessary.

zoakesAStuff commented on src/udp_interface.cpp:
```diff
@@ -16,97 +19,79 @@ UDPInterface::UDPInterface() :
 
   boost::system::error_code ec;
-  bytes_read = socket_.receive_from(boost::asio::buffer(msg, buf_size), sender_endpoint_, 0, ec);
+  msg->assign(10000, 0);
```
See comment in tcp_interface

JWhitleyWork replied:
Same.

zoakesAStuff commented on src/udp_interface.cpp:
```diff
@@ -16,97 +19,79 @@ UDPInterface::UDPInterface() :
 }
 
-// Default destructor.
-UDPInterface::~UDPInterface()
```
+1 for deleting empty destructor

zoakesAStuff commented on test/network_interface_test.cpp:
```diff
@@ -5,50 +5,51 @@
   // 123.4567 as binary float
-  uint8_t bytes[4] = { 0x42, 0xF6, 0xE9, 0xD5};
-  ASSERT_EQ(123.45670318603515625, AS::Network::read_be<float>(bytes, 4, 0));
+  std::vector<uint8_t> bytes = { 0x42, 0xF6, 0xE9, 0xD5};
```
Doesn't matter at all, but the brace spacing is inconsistent in all these initializers

Commit:
Consistent braced initializers.

Commit:
Removing apparently useless check in write_le.

JWhitleyWork commented:
Closing this pull request until I figure out how to address comments in https://github.com/astuff/network_interface/pull/25/files#r268292600

Commit:
Removing async calls from TCP. Allocating buffers with reserve.

JWhitleyWork commented:
So, for now, I modified the read calls to all be synchronous. We can add asynchronous calls but in a more responsible way, and at a later date. This was just intended to be clean-up.

zoakesAStuff commented on include/network_interface/network_utils.h:
```diff
@@ -8,21 +8,19 @@
-};
 
 inline bool system_is_big_endian()
 {
```
As I found out the other day, this is relying on undefined behavior in C++. I think the correct way to do this would be with a call to std::reinterpret_cast.

JWhitleyWork replied:
You're correct. In order for the compiler not to balk at it, I had to do a `reinterpret_cast` on a pointer to the type but it does work correctly (as shown by the tests). Should be good now.

zoakesAStuff replied:
All the tests pass, but the CircleCI machines run x86_64 same as our dev machines. Since x64 is what this is most likely to be run on that's probably good. However a lot of ARM machines run big-endian so I'd like to test on a BE system if we plan on supporting any other architectures.

JWhitleyWork replied:
There is a problem running non-x86 images on CircleCI. See https://support.circleci.com/hc/en-us/articles/360018651374-Job-completes-after-Spin-up-Environment-for-non-x86-based-Docker-images-. I tested this and they fail exactly as described. I've asked for more information at https://ideas.circleci.com/ideas/CCI-I-900 but I doubt I'll see a fix anytime soon. I'll test this locally to see if I can get it to run.

zoakesAStuff replied:
@JWhitleyAStuff Did you have any success with testing locally?

JWhitleyWork replied:
I haven't had time, so no. Also, I couldn't get the stock cross-platform `ros` Docker images to run on my machine (even after doing the qemu instructions to get them going) though I could run the cross-platform `ubuntu` images.

srustanAStuff replied:
Have you since had time to test on a local machine?

JWhitleyWork replied:
@srustanAStuff - I finally just got it installed and running. The tests passed correctly. I'm going to try to set up CI to do this test as well. Please hold on approval until I have that added to this PR.

JWhitleyWork replied:
Holy carp, it worked! Now testing (and passing) against `armhf` on CircleCI.

Commit:
Changing to reinterpret_cast instead of non-portable solution.

JWhitleyWork commented:
Ping @zoakesAStuff

Commit:
CI: Add first run at armhf build.

Commit:
CI: Machine executors require sudo.

Commit:
CI: Adding Melodic armhf build.

Commit:
Used wrong type-trait qualifier.

JWhitleyWork commented:
@Daniel-Stanek / @samrustan / @jurlaubAStuff - Can you guys review the change I made in 3732ce3? I found while trying to build ibeo_core against this new version that I used the wrong `type_trait` specifier.

Daniel-Stanek commented:
Took a closer look at the entire network utils file instead of just the changes.
Lines 30/46, 82/96: why is valueOffset a uint32_t that's not cast to T?  Shouldn't it be the same as factor?

Lines 53, 80, 86, 103: "unsigned int" should be uint32_t

Line 58-75: I don't quite get what write_le is trying to do.  It looks like it just returns an empty array if the input is a floating point.  Then how does the condition on 68 work, obviously it will equal 0 and stop once mask has shifted 0xFF all the way past the upper most byte, but won't it also stop as soon as it hits a byte that's zero in source?  Finally on 70 doesn't casting to uint8_t take the least most byte, you're shifting the mask up but never shifting the result down?  Actually it looks like these were fixed in write_be.

Line 39 vs. 89: it looks like offset is zero based in one place and not in the other

Commit:
Replacing unsigned int with uint32_t.

Commit:
Changing valueOffset to float and casting in util functions.

JWhitleyWork commented:
> Took a closer look at the entire network utils file instead of just the changes.
> Lines 30/46, 82/96: why is valueOffset a uint32_t that's not cast to T? Shouldn't it be the same as factor?

You're right, it should also be a float that is cast to T. I didn't realize that the offset could also be a float when I initially wrote this. Changed in c926482.

> Lines 53, 80, 86, 103: "unsigned int" should be uint32_t

Fixed in bf18443.

> Line 58-75: I don't quite get what write_le is trying to do. It looks like it just returns an empty array if the input is a floating point. Then how does the condition on 68 work, obviously it will equal 0 and stop once mask has shifted 0xFF all the way past the upper most byte, but won't it also stop as soon as it hits a byte that's zero in source? Finally on 70 doesn't casting to uint8_t take the least most byte, you're shifting the mask up but never shifting the result down? Actually it looks like these were fixed in write_be.

These were pretty naive implementations. I'm working on new ones now. Thanks for pointing this out.

> Line 39 vs. 89: it looks like offset is zero based in one place and not in the other

That's because in line 39, the for loop counts from `sizeof(T)` down to 1 and in line 89, it counts from 0 to `sizeof(T) - 1`.

Commit:
New implementation of write_le and write_be. Tests for both.

srustanAStuff commented on include/network_interface/network_utils.h:
```diff
@@ -56,24 +56,48 @@ T read_le(const std::vector<uint8_t>& bufArray,
+  uint32_t shift = 0;
 
-  while ((*source & mask) > 0)
+  while (shift < (8 * sizeof(T)))
```
8 meaning 8 bits or bytes.  It seems trivial, but using #defines for these numbers is helpful for quick comprehension.

JWhitleyWork replied:
@srustanAStuff In this context, it can only mean bits. `sizeof()` returns a number of bytes and `shift` represents a number of bits by which to shift a value. To convert from the return value of `sizeof()` to a number of bits to shift, you have to multiply by 8. Replacing the number 8 with any `#define` would be more verbose than necessary here (e.g. `BITS_IN_BYTE`). Readability is important but a certain level of code comprehension is assumed.

srustanAStuff commented on include/network_interface/network_utils.h:
```diff
@@ -105,22 +129,19 @@ T read_be(const std::vector<uint8_t>& bufArray,
   return read_be<T>(bufArray, offset, 1.0, 0);
```
what is 1.0 and 0?

JWhitleyWork replied:
If you look at the implementation with 4 arguments at line 103 above, they are the `factor` and `valueOffset`. This version of `read_be` only takes 2 arguments (the array and the byte offset of the value to be read) so it assumes you don't want a factor or value offset so it sets them to innocuous values.

JWhitleyWork commented:
@Daniel-Stanek - Should be fixed.

jurlaubAStuff reviewed (dismissed):
Is there not a library that addresses le & be reading and writing? was it simpler to write them here?
I am also not a fan of 'magic numbers' - my preference would be to at least have a comment even if it should be obvious.

JWhitleyWork commented:
> my preference would be to at least have a comment even if it should be obvious.

I can comment the code better. I think that may have been what @samrustan was ultimately getting at.

Commit:
Commenting code.

JWhitleyWork commented:
@jurlaubAStuff - ~~There are POSIX functions which do this. If we wanted to tie this to POSIX-style systems (basically Linux, Unix, and BSD), then we could use those. However, they are C-only, their header locations are system-dependent, and once ROS2 rolls around (with support for Windows and Mac OS), they will no longer be an option.~~

Nevermind again. These are associated with determining (or converting to/from) the storage order on a particular machine. We are converting within a data-stream. I think this implementation works fine.

~~https://github.com/steinwurf/endian might be another option.~~
Nevermind. Above uses the "undefined behavior" example. https://github.com/steinwurf/endian/blob/master/src/endian/is_big_endian.hpp

Commit:
Allowing read functions to take multiple containers.

Commit:
Moving parse_tuple from ibeo_core.

JWhitleyWork commented:
@jurlaubAStuff / @Daniel-Stanek / @srustanAStuff / @rgunjala93 - Last change, guys, I *swear*. I had to move the `parse_tuple` function here from `ibeo_core` to avoid a linker error relating to `network_utils`. I also modified the `read_x` functions so they could work on any type of container that supports the `[]` operator. This allows the use of `deque` in ibeo_lux and ibeo_scala, which are significantly more efficient in their use cases. Sorry again for the multiple reviews.

Commit:
Had to add offset to parse_tuple.
